#**Case Study: Face Care Products in Online E-commerce Platform Shopee**
A total of 120 face care products from Shopee were extracted on the 7th December 2021. This project covers the typical data or text preprocessing applied before data visualization. The objective of this study is to analyse the distribution of data related to face care products and to identify the best selling face care product in Shopee (**MAIN FOCUS OF ASSIGNMENT 2: Data Visualization**)

---
**Name:** Tang Jia Hui <br>
**Matric Number:** A176297<br>

---
Note: There are 5 CSV files and 2 HTML files in the A176297_TC3213_CSV.zip file. You may download the zip file at: https://drive.google.com/file/d/1n0zfzKNZpFYPnXx9qVshLPW7aTkAIsKV/view?usp=sharing

1. **face_care_product_link.csv** is the csv file that stores all 120 links relative to the face care products. It is obtained in Part 1 and will be used throughout the rest of the project.
2. **facecare_df.csv** is the csv file that stores all 120 different face care product attributes (uncleaned) scraped from Shopee. It is obtained from Part 2-6 of this project and is used in Part 8 of this project.
3. **a176297_cleaned_face_care.csv** is the csv file that stores the cleaned but untransformed data respective to the remaining face care products, without changing the categorical data to quantitative data.
4. **a176297_cleaned_transform_face_care.csv** is the final csv file that stores the cleaned and transformed data respective to the remaining face care products scraped.
5. **a176297_cleaned_transform_one_hot_face_care.csv** is the final csv file that stores the cleaned and transformed data respective to the remaining face care products scraped. Besides, all categorical data are converted to one hot encodings
6. **a176297_norm_output.html** and **a176297_norm_output.html** are the data profiles generated.

---

**ETL PROCESS:**<br>
    a. Extraction: PART 1-7 <br>
    b. Transformation: PART 8 <br>
    c. Load: PART 9<br>
    d. Data Visualization / EDA: PART 10

##**Step 1: Import Library**
Import the necessary libraries to fulfill the work of this project

In [ ]:
# Install and update library
!pip install selenium
!apt-get update # to update ubuntu to correctly run apt install
!apt install chromium-chromedriver
# !pip install urllib3
!pip install folium
!pip install albumentations
!pip install -q gwpy
# %%capture 

  Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.25.11
    Uninstalling urllib3-1.25.11:
      Successfully uninstalled urllib3-1.25.11
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests 2.23.0 requires urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1, but you have urllib3 1.26.7 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.
Get:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease
Ign:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:5 http://ppa.launchpad.net/c2d4u.team/c2d4

In [ ]:
from selenium import webdriver
from selenium.webdriver.common.by import By
# !cp /usr/lib/chromium-browser/chromedriver /usr/bin
import sys
import pandas as pd
import plotly.express as px

#**PART 1: SELENIUM EXTRACT FACE CARE PRODUCT LINKS**
Selenium, an open-source umbrella project consisting a range of tools and libraries used for web browser automation and web scraping will be used as the tool to scrape product (face care products) links in order to ease the scraping process of face care products information in the latter section of this project.



##**Step 1: Extract Anchor Link of Products**
A function named get_product_links is defined to extract the links of each respective face care products in Shopee. Then, the function is called to start the extraction process and the extracted link is stored in a list named as product_link. A total of 120 face care product links are extracted.

In [ ]:
%%capture
# define function to extract product links
def get_product_links(page_num, product_link_lst):
    # Find link that access product page
    url = "https://shopee.com.my/mall/search?brands=1802087%2C1801327%2C1006592%2C1802669%2C1150423%2C1007876%2C1009373%2C1801694%2C1802827%2C1800923%2C1800631%2C1648458%2C1802456%2C1006035&keyword=face%20care&noCorrection=true&page=" + str(page_num)
    # url = "https://shopee.com.my/mall/search?brands=1802087%2C1801327%2C1006592%2C1802669%2C1150423%2C1007876%2C1009373%2C1801694%2C1802827%2C1800923%2C1800631&keyword=face%20care&noCorrection=true&page=" + str(page_num)
    #Define Chrome WebDriver
    sys.path.insert(0,'/usr/lib/chromium-browser/chromedriver')
    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver', options=chrome_options)
    # get url
    wd.get(url)
    # impose pause
    wd.implicitly_wait(10)
    container = wd.find_element_by_xpath('/html/body/div[1]/div/div[3]/div/div[2]/div').find_elements_by_tag_name('a')
    for link in container:
        product_link_lst.append(link.get_attribute('href'))
    wd.quit()

# Call function get_product_link
product_link = list()
for page_num in range(0, 8):
    get_product_links(page_num, product_link)


In [ ]:
len(product_link)

120

In [ ]:
product_link

['https://shopee.com.my/face-factory-Face-Factory-Skin-Booster-Wireless-i.432254454.3688068941?sp_atk=77e4f160-e6db-4855-9e40-f962b954353f',
 'https://shopee.com.my/LAIKOU-Sakura-Toner-Face-Serum-Reduce-Spots-Wrinkles-Skin-Care-Set-6pcs-i.132668829.8118106855?sp_atk=82c86420-3a82-49c9-bac8-27bb719f85ad',
 'https://shopee.com.my/Garnier-Light-Complete-Brightening-Serum-Mask-For-Dull-Skin-with-Dark-Spots-Brightening-Whitening-(5-Pack)-i.55790034.6741347404?sp_atk=062ad0b9-c3e3-4377-95a8-c152babe514a',
 'https://shopee.com.my/CkeyiN-Electric-Facial-Massage-V-Line-Cheek-Lift-Up-Tools-for-Women-Double-Chin-Facial-Weight-Loss-for-Face-Care-MR512-i.43706140.7844157691?sp_atk=08275109-cb9c-442a-bafa-a7b2d9cdbaf3',
 'https://shopee.com.my/NIVEA-Face-Care-Cleanser-Hokkaido-Rose-Whip-(100ml)-Face-Wash-Make-up-remover-Rose-Natural-Ingredients-i.25459176.5009744064?sp_atk=e6633f31-a4b4-424b-9e10-1c367a02d960',
 'https://shopee.com.my/LAIKOU-Japan-Sakura-Cleanser-Whitening-Face-Serum-Reduce-Spots-Wr

In [ ]:
len(product_link)

120

In [ ]:
len(set(product_link))

120

##**Step 2: Save product links in a CSV file**
The product links are saved to a CSV file to work with the extracted links in the following section.

In [ ]:
link_dict = {'product_link': product_link}
df_link = pd.DataFrame(link_dict)
df_link.to_csv('face_care_product_link.csv')

#**PART 2: EXTRACT FACE CARE CATEGORY WITH SELENIUM**
The extraction of the category of face care products is performed with selenium as well. Before the extraction process begins, the csv file saved in the previous section (Part 1) is read and the links are restored as a list, called product_link.

In [ ]:
facecare_link_df = pd.read_csv('/content/face_care_product_link.csv', index_col='Unnamed: 0').values.tolist()
facecare_link_lst = [link[0] for link in facecare_link_df]
facecare_link_lst

['https://shopee.com.my/face-factory-Face-Factory-Skin-Booster-Wireless-i.432254454.3688068941?sp_atk=77e4f160-e6db-4855-9e40-f962b954353f',
 'https://shopee.com.my/LAIKOU-Sakura-Toner-Face-Serum-Reduce-Spots-Wrinkles-Skin-Care-Set-6pcs-i.132668829.8118106855?sp_atk=82c86420-3a82-49c9-bac8-27bb719f85ad',
 'https://shopee.com.my/Garnier-Light-Complete-Brightening-Serum-Mask-For-Dull-Skin-with-Dark-Spots-Brightening-Whitening-(5-Pack)-i.55790034.6741347404?sp_atk=062ad0b9-c3e3-4377-95a8-c152babe514a',
 'https://shopee.com.my/CkeyiN-Electric-Facial-Massage-V-Line-Cheek-Lift-Up-Tools-for-Women-Double-Chin-Facial-Weight-Loss-for-Face-Care-MR512-i.43706140.7844157691?sp_atk=08275109-cb9c-442a-bafa-a7b2d9cdbaf3',
 'https://shopee.com.my/NIVEA-Face-Care-Cleanser-Hokkaido-Rose-Whip-(100ml)-Face-Wash-Make-up-remover-Rose-Natural-Ingredients-i.25459176.5009744064?sp_atk=e6633f31-a4b4-424b-9e10-1c367a02d960',
 'https://shopee.com.my/LAIKOU-Japan-Sakura-Cleanser-Whitening-Face-Serum-Reduce-Spots-Wr

##**Step 1: Define Function to Extract Face Care Product Category**
The function named get_category is used to extract the category for each face care product.

In [ ]:
# /html/body/div[1]/div/div[2]/div[2]/div[2]/div[4]/div[2]/div[1]/div[1]/div[1]/div[2]/div[1]/div/span[2]
def get_category(links, facecare_category):
    # webdriver is imported in Part 1 Step 1
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver',options=options)
    wd.get(links)
    wd.implicitly_wait(10)
    facecare_sub = wd.find_elements(By.CLASS_NAME, '_3QRNmL')
    facecare_sub[2].get_attribute('innerHTML')
    for index, value in enumerate(facecare_sub):
        if (index == 2) and (value.get_attribute('innerHTML') != 'For Face'):
            facecare_category.append(value.get_attribute('innerHTML'))
        elif (index == 2) and (value.get_attribute('innerHTML') == 'For Face'):
            facecare_category.append("Men's Grooming")
    wd.quit()

##**Step 2: Extraction of all face care products' category**
The function get_category is called and the category for each face care product is saved in variable facecare_category_lst are listed. Before saving the extract category to a dataframe called face_df, the unique category of face care products will be stored in variable named as unique_category and its values are listed later.

In [ ]:
facecare_category_lst = []
for i in facecare_link_lst:
    get_category(i, facecare_category_lst)
print('Scraping Face Care Product Category -- Done!')

Scraping Face Care Product Category -- Done!


In [ ]:
len(facecare_category_lst)

120

In [ ]:
unique_category = set(facecare_category_lst)
unique_category

{'Acne Treatment',
 'Body Lotion &amp; Others',
 'Body Scrubs, Exfoliant &amp; Peel',
 'Body Sunscreen &amp; After Sun',
 'Body Wash &amp; Soap',
 'Cleanser',
 'Cotton Balls, Pads &amp; Buds',
 'Essence &amp; Serum',
 'Eye Cream &amp; Serum',
 'Face Mask &amp; Packs',
 'Face Scrub &amp; Peel',
 'Facial Tools',
 'Foundation',
 'Gift Set',
 'Hair &amp; Scalp Treatments',
 'Hair Styling',
 'Health Accessories',
 'Make Up Remover',
 'Masks, Lotion, Cream &amp; Scrubs',
 "Men's Grooming",
 'Moisturizer &amp; Cream',
 'Slimmers &amp; Massagers',
 'Sunscreen &amp; Aftersun',
 'Toner &amp; Mists'}

In [ ]:
face_df = pd.DataFrame(facecare_category_lst, columns=['facecare_category'])
face_df

facecare_category
0    Slimmers &amp; Massagers
1           Toner &amp; Mists
2       Face Mask &amp; Packs
3    Slimmers &amp; Massagers
4                    Cleanser
..                        ...
115  Sunscreen &amp; Aftersun
116  Sunscreen &amp; Aftersun
117            Men's Grooming
118      Body Wash &amp; Soap
119                  Gift Set

[120 rows x 1 columns]

#**PART 3: EXTRACT FACE CARE PRODUCT BRAND WITH SELENIUM**
The extraction of face care products brand utilize selenium and the list called facecare_link_lst.

##**Step 1: Define Function to Extract Face Care Product Brand**
The function named get_brand is used to extract the brand for each face care product.

In [ ]:
def get_brand(links, face_brand_lst):
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver',options=options)
    wd.get(links)
    wd.implicitly_wait(10)
    face_sub = wd.find_elements(By.CLASS_NAME, '_3Qy6bH')
    if(len(face_sub)>0):
        face_brand_lst.append(face_sub[0].get_attribute('innerHTML'))
    else:
        face_sub =wd.find_elements(By.CLASS_NAME, '_3uf2ae')
        face_brand_lst.append(face_sub[0].get_attribute('innerHTML'))
        # print(face_sub[0].get_attribute('innerHTML'))
    wd.quit()

##**Step 2: Extraction of all face care products' brand**
The function get_brand is called and the brand for each face care product is saved in variable face_brand_lst are listed. Before saving the extract brand to a dataframe called face_df, the unique brand of face care products will be stored in variable named as unique_brand and its values are listed later.

In [ ]:
face_brand_lst = []
for i in facecare_link_lst:
    get_brand(i, face_brand_lst)
print('Scraping Face Care Brand -- Done!')

Scraping Face Care Brand -- Done!


In [ ]:
face_brand_lst

['Face Factory',
 'Laikou',
 'Garnier',
 'Ckeyin',
 'Nivea',
 'Laikou',
 'Nivea',
 'Nivea',
 'Lanbena',
 'Lanbena',
 'Lanbena',
 'Hada Labo',
 'Nivea',
 'Nivea',
 'Nivea',
 'Himalaya',
 'Nivea',
 'VT Cosmetics',
 'Nivea',
 'Eaoron',
 'Lanbena',
 'Vibrant Glamour',
 'Nivea',
 'Laikou',
 'Nivea',
 'Vibrant Glamour',
 'Lanbena',
 'The Face Shop',
 'Nivea',
 'Nivea',
 'Breylee',
 'Nivea',
 'Nivea',
 'sunniesfaceofficial.os',
 'Breylee',
 'Nivea',
 'Vibrant Glamour',
 'Himalaya',
 'Hada Labo',
 'Laikou',
 'Hada Labo',
 'Huxley',
 'Lanbena',
 'Vibrant Glamour',
 'Hada Labo',
 'Laikou',
 'Wis',
 'telekungrawdah.os',
 'Simply K',
 "LIYAL'AN",
 'Focallure',
 'Himalaya',
 'Hada Labo',
 'Nivea',
 'Himalaya',
 'Bifesta',
 'Vibrant Glamour',
 'Huxley',
 'Lanbena',
 'Nivea',
 'Nivea',
 'Laikou',
 'Vibrant Glamour',
 'Lanbena',
 'Himalaya',
 'Laikou',
 'Nivea',
 'The Face Shop',
 'Laikou',
 'The Face Shop',
 'Laikou',
 'Nivea',
 'Curel',
 'Nivea',
 'The Face Shop',
 'Himalaya',
 'Huxley',
 'Nivea',
 

In [ ]:
len(face_brand_lst)

120

In [ ]:
unique_brands = set(face_brand_lst)
unique_brands

{'Bifesta',
 'Breylee',
 'Ckeyin',
 'Curel',
 'Dalton Marine Cosmetics',
 'Eaoron',
 'Face Factory',
 'Focallure',
 'Garnier',
 'Hada Labo',
 'Himalaya',
 'Huxley',
 'KingdomCares',
 "LIYAL'AN",
 'Laikou',
 'Lanbena',
 'Nivea',
 'SNP',
 'Sebamed',
 'Simply K',
 'The Face Shop',
 'VT Cosmetics',
 'Vibrant Glamour',
 'Wis',
 'sunniesfaceofficial.os',
 'telekungrawdah.os'}

In [ ]:
face_df['facecare_brand'] = pd.DataFrame(face_brand_lst, columns=['facecare_brand'])
face_df

facecare_category facecare_brand
0    Slimmers &amp; Massagers   Face Factory
1           Toner &amp; Mists         Laikou
2       Face Mask &amp; Packs        Garnier
3    Slimmers &amp; Massagers         Ckeyin
4                    Cleanser          Nivea
..                        ...            ...
115  Sunscreen &amp; Aftersun            SNP
116  Sunscreen &amp; Aftersun            SNP
117            Men's Grooming         Laikou
118      Body Wash &amp; Soap  The Face Shop
119                  Gift Set         Huxley

[120 rows x 2 columns]

#**PART 4: EXTRACT FACE CARE PRODUCT NAME WITH SELENIUM**
The extraction of face care products brand utilize selenium and the list called facecare_link_lst.

##**Step 1: Define Function to Extract Face Care Product Name**
The function named get_name is used to extract the name for each face care product.

In [ ]:
def get_name(links, face_name_lst):
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver',options=options)
    wd.get(links)
    wd.implicitly_wait(10)
    face_name_sub = wd.find_elements(By.XPATH, '/html/body/div[1]/div/div[2]/div[2]/div/div[2]/div[3]/div/div[1]/span')
    face_name_lst.append(face_name_sub[0].get_attribute('innerHTML'))
    wd.quit()

##**Step 2: Extraction of all face care products' name**
The function get_name is called and the name for each face care product is saved in variable face_name_lst are listed. The extract name will be saved to the dataframe face_df.

In [ ]:
face_name_lst = []
for i in facecare_link_lst:
    get_name(i, face_name_lst)
print('Scraping Face Care Product Name -- Done!')

Scraping Face Care Product Name -- Done!


In [ ]:
face_name_lst

['face factory Face Factory Skin Booster Wireless',
 'LAIKOU Sakura Toner Face Serum Reduce Spots Wrinkles Skin Care Set 6pcs',
 'Garnier Light Complete Brightening Serum Mask For Dull Skin with Dark Spots Brightening/Whitening (5 Pack)',
 'CkeyiN Electric Facial Massage V-Line Cheek Lift Up Tools for Women Double Chin Facial Weight Loss for Face Care MR512',
 'NIVEA Face Care Cleanser - Hokkaido Rose Whip (100ml) | Face Wash | Make up remover | Rose, Natural Ingredients',
 'LAIKOU Japan Sakura Cleanser Whitening Face Serum Reduce Spots Wrinkles Skin Care Set',
 'NIVEA Face Care Cleanser - Glow (Raspberry Scrub)(75ml)',
 'NIVEA Face Care Cleanser - Glow - Raspeberry Scrub (75ml) | Face Scrub | Natural Ingredients | Exfoliating Face Wash',
 'LANBENA Blackhead Remover Face Mask Peel-Off Acne Treatment Deep Cleansing Shrink Pore  Natural Bamboo Charcoal Skin Care',
 'LANBENA Facial Mask Skin Care Whitening Hydrating Anti-Wrinkle',
 'LANBENA Vitamin C Whitening Serum Hyaluronic Acid Face C

In [ ]:
len(face_name_lst)

120

In [ ]:
face_df['facecare_name'] = pd.DataFrame(face_name_lst, columns=['facecare_name'])
face_df

facecare_category  ...                                      facecare_name
0    Slimmers &amp; Massagers  ...    face factory Face Factory Skin Booster Wireless
1           Toner &amp; Mists  ...  LAIKOU Sakura Toner Face Serum Reduce Spots Wr...
2       Face Mask &amp; Packs  ...  Garnier Light Complete Brightening Serum Mask ...
3    Slimmers &amp; Massagers  ...  CkeyiN Electric Facial Massage V-Line Cheek Li...
4                    Cleanser  ...  NIVEA Face Care Cleanser - Hokkaido Rose Whip ...
..                        ...  ...                                                ...
115  Sunscreen &amp; Aftersun  ...  SNP : UV Perfect Air-Cool Sun Gel + Air Safe S...
116  Sunscreen &amp; Aftersun  ...  SNP : Uv Perfect Air-Lock Sun Fluid + Air Safe...
117            Men's Grooming  ...  LAIKOU Men  Face Skin Care Set Face Toner Loti...
118      Body Wash &amp; Soap  ...               Dr.Belmeur Mild Derma Acne Body Wash
119                  Gift Set  ...  Huxley Official Conditioning Skincare Set [ Se...

[120 rows x 3 columns]

#**PART 5: EXTRACT FACE CARE PRODUCT PRICE WITH SELENIUM**
The extraction of face care products price utilize selenium and the list called facecare_link_lst.

##**Step 1: Define Function to Extract Face Care Product Price**
The function named get_price is used to extract the price for each face care product.

In [ ]:
def get_price(links, face_price_lst):
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver',options=options)
    wd.get(links)
    wd.implicitly_wait(5)
    face_price_sub = wd.find_elements(By.CLASS_NAME, 'Ybrg9j')
    if face_price_sub[0].get_attribute('innerHTML') != " ":
        face_price_lst.append(face_price_sub[0].get_attribute('innerHTML'))
        # print(face_price_sub[0].get_attribute('innerHTML'))
        # return face_price_sub[0].get_attribute('innerHTML')
    else:
        face_price_sub = wd.find_elements(By.CLASS_NAME, '_2MaBXe')
        # print(face_price_sub[0].get_attribute('innerHTML'))
        face_price_lst.append(face_price_sub[0].get_attribute('innerHTML'))
        # print(face_price_sub[0].get_attribute('innerHTML'))
    
    wd.quit()

##**Step 2: Extraction of all face care products' price**
The function get_price is called and the price for each face care product is saved in variable face_price_lst are listed. The extract price will be saved to the dataframe face_df.

In [ ]:
face_price_lst = []
for i in facecare_link_lst:
    get_price(i, face_price_lst)
print('Scraping Face Care Product Price -- Done!')

Scraping Face Care Product Price -- Done!


In [ ]:
len(face_price_lst)

120

In [ ]:
face_price_lst

['RM230.00',
 'RM36.99',
 'RM38.25',
 'RM46.00 - RM48.59',
 'RM16.72',
 'RM11.66 - RM23.32',
 'RM16.56',
 '',
 'RM3.59',
 'RM2.00 - RM9.00',
 'RM7.29',
 'RM22.69',
 'RM15.92',
 'RM15.92',
 'RM20.90',
 'RM12.59',
 'RM15.92',
 'RM79.90',
 'RM41.79',
 'RM40.00',
 'RM9.37',
 'RM9.90',
 'RM15.92',
 'RM5.09',
 'RM29.52',
 'RM7.90',
 'RM11.25',
 '',
 'RM16.72',
 'RM20.90',
 'RM8.75',
 'RM22.76',
 'RM15.92',
 'RM36.54',
 'RM10.70',
 'RM15.12',
 'RM23.75',
 'RM9.68',
 'RM50.26',
 'RM7.98',
 'RM30.50',
 'RM79.90',
 'RM12.80',
 'RM23.70 - RM49.70',
 'RM48.86',
 'RM3.99 - RM6.99',
 'RM49.50 - RM79.00',
 'RM17.00',
 'RM11.90',
 'RM15.98',
 '',
 'RM10.05',
 'RM48.86',
 'RM42.42',
 'RM29.31',
 'RM26.32',
 'RM29.40',
 'RM159.80 - RM189.90',
 'RM7.40',
 'RM42.42',
 'RM15.92',
 'RM7.79',
 'RM3.70',
 'RM2.95',
 'RM18.87',
 'RM2.61',
 'RM17.52',
 'RM34.00',
 'RM5.73',
 'RM98.90',
 'RM5.90',
 'RM16.72',
 'RM75.42',
 'RM35.42',
 'RM47.90',
 'RM15.50',
 'RM120.00',
 'RM35.67',
 'RM70.50',
 'RM29.31',
 'RM99.

In [ ]:
len([i for i,v in enumerate(face_price_lst) if v == ''])

4

In [ ]:
face_df['facecare_price'] = pd.DataFrame(face_price_lst, columns=['facecare_price'])
face_df

facecare_category  ...     facecare_price
0    Slimmers &amp; Massagers  ...           RM230.00
1           Toner &amp; Mists  ...            RM36.99
2       Face Mask &amp; Packs  ...            RM38.25
3    Slimmers &amp; Massagers  ...  RM46.00 - RM48.59
4                    Cleanser  ...            RM16.72
..                        ...  ...                ...
115  Sunscreen &amp; Aftersun  ...            RM71.25
116  Sunscreen &amp; Aftersun  ...            RM71.25
117            Men's Grooming  ...            RM49.50
118      Body Wash &amp; Soap  ...            RM95.90
119                  Gift Set  ...           RM259.90

[120 rows x 4 columns]

#**PART 6: EXTRACT FACE CARE PRODUCT TOTAL SOLD WITH SELENIUM**
The extraction of face care products sales (total sold) utilize selenium and the list called facecare_link_lst.<br>
**Note: The total sold means the total quantity of the product purchased by customers**

##**Step 1: Define Function to Extract Face Care Product Total Sold**
The function named get_total_sold is used to extract the total sold for each face care product.

In [ ]:
def get_total_sold(links, face_price_lst):
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    wd = webdriver.Chrome('chromedriver',options=options)
    wd.get(links)
    wd.implicitly_wait(5)
    face_price_sub = wd.find_elements(By.CLASS_NAME, 'aca9MM')
    face_price_lst.append(face_price_sub[0].get_attribute('innerHTML'))
    wd.quit()

##**Step 2: Extraction of all face care products' price**
The function get_total_sold is called and the total sold for each face care product is saved in variable face_sold_lst are listed. The extract price will be saved to the dataframe face_df.

In [ ]:
face_sold_lst = []
for i in facecare_link_lst:
    get_total_sold(i, face_sold_lst)
print('Scraping Face Care Product Sales -- Done!')

Scraping Face Care Product Sales -- Done!


In [ ]:
face_sold_lst

['23',
 '258',
 '1.1k',
 '715',
 '2.9k',
 '2.6k',
 '239',
 '2.3k',
 '1.3k',
 '5.3k',
 '7.6k',
 '15.6k',
 '1.6k',
 '906',
 '91',
 '76',
 '761',
 '419',
 '72',
 '1.9k',
 '147',
 '1.7k',
 '1.5k',
 '259',
 '3.6k',
 '604',
 '533',
 '86',
 '565',
 '43',
 '6.6k',
 '3.1k',
 '2.5k',
 '122',
 '2.1k',
 '445',
 '64',
 '1.2k',
 '7',
 '286',
 '60',
 '19',
 '282',
 '7',
 '280',
 '4.2k',
 '99',
 '207',
 '9',
 '0',
 '2.2k',
 '1.3k',
 '236',
 '147',
 '286',
 '406',
 '29',
 '13',
 '278',
 '168',
 '18',
 '904',
 '180',
 '132',
 '138',
 '272',
 '516',
 '54',
 '116',
 '37',
 '6',
 '403',
 '5',
 '1',
 '48',
 '601',
 '11',
 '13',
 '19',
 '404',
 '0',
 '9',
 '13',
 '652',
 '1',
 '761',
 '0',
 '76',
 '10',
 '18',
 '38',
 '3',
 '1',
 '5',
 '1',
 '11',
 '0',
 '144',
 '17',
 '0',
 '30',
 '132',
 '17',
 '0',
 '0',
 '5',
 '1',
 '361',
 '0',
 '7',
 '0',
 '3',
 '0',
 '0',
 '24',
 '4',
 '1',
 '2',
 '0',
 '0']

In [ ]:
len(face_sold_lst)

120

In [ ]:
face_df['facecare_total_sold'] = pd.DataFrame(face_sold_lst, columns=['facecare_total_sold'])
face_df

facecare_category  ... facecare_total_sold
0    Slimmers &amp; Massagers  ...                  23
1           Toner &amp; Mists  ...                 258
2       Face Mask &amp; Packs  ...                1.1k
3    Slimmers &amp; Massagers  ...                 715
4                    Cleanser  ...                2.9k
..                        ...  ...                 ...
115  Sunscreen &amp; Aftersun  ...                   4
116  Sunscreen &amp; Aftersun  ...                   1
117            Men's Grooming  ...                   2
118      Body Wash &amp; Soap  ...                   0
119                  Gift Set  ...                   0

[120 rows x 5 columns]

#**PART 7: SAVE EXTRACTED DATA TO CSV File**
The final data frame that has the uncleaned data about the face care products is performed and will be used in PART 8: DATA CLEANING

In [ ]:
face_df.to_csv('facecare_df.csv')

#**PART 8: DATA CLEANING AND TRANSFORMATION**
From Part 2 to 6, some of the data extracted from Shopee is irrelevant to the case study. Referring to Part 2 - Step 2, the unique_style showed 24 different categories and some of which are irrelevant to face care products such as Cotton Balls, Pads &amp; Buds, hair styling and more. Then, there are categories of the same type but were listed as different categories, namely the *body sunscreen and after sun* and *sunscreen and aftersun*. Besides, some of the face care product price weren't extracted, hence indicating the presence of null values in the cell of dataframe. Moreover, some of the price extracted consisted of 2 values such as RM46.00 - RM48.59 in a single cell and the letter **RM** in the cells for that column. In addition, the total sold of face care product consists of **k** which represents the number in thousands. The mentioned problems will be addressed and explained in the following section. <br>
**Note:** The products under the 'body sunscreen and after sun' category actually refers to sunscreen for faces, instead of body

In [ ]:
import pandas as pd
import re
import statistics
face_care = pd.read_csv('/content/facecare_df.csv', index_col='Unnamed: 0')
face_care.head()

facecare_category  ... facecare_total_sold
0  Slimmers &amp; Massagers  ...                  23
1         Toner &amp; Mists  ...                 258
2     Face Mask &amp; Packs  ...                1.1k
3  Slimmers &amp; Massagers  ...                 715
4                  Cleanser  ...                2.9k

[5 rows x 5 columns]

In [ ]:
face_care.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120 entries, 0 to 119
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   facecare_category    120 non-null    object
 1   facecare_brand       120 non-null    object
 2   facecare_name        120 non-null    object
 3   facecare_price       116 non-null    object
 4   facecare_total_sold  120 non-null    object
dtypes: object(5)
memory usage: 5.6+ KB


In [ ]:
face_care['facecare_category'].unique()

array(['Slimmers &amp; Massagers', 'Toner &amp; Mists',
       'Face Mask &amp; Packs', 'Cleanser', 'Essence &amp; Serum',
       'Face Scrub &amp; Peel', "Men's Grooming",
       'Moisturizer &amp; Cream', 'Masks, Lotion, Cream &amp; Scrubs',
       'Make Up Remover', 'Acne Treatment',
       'Body Sunscreen &amp; After Sun', 'Health Accessories',
       'Cotton Balls, Pads &amp; Buds', 'Gift Set',
       'Body Scrubs, Exfoliant &amp; Peel', 'Hair Styling',
       'Sunscreen &amp; Aftersun', 'Body Lotion &amp; Others',
       'Foundation', 'Body Wash &amp; Soap',
       'Hair &amp; Scalp Treatments', 'Eye Cream &amp; Serum',
       'Facial Tools'], dtype=object)

In [ ]:
len(face_care['facecare_category'].unique())

24

In [ ]:
face_care['facecare_brand'].unique()

array(['Face Factory', 'Laikou', 'Garnier', 'Ckeyin', 'Nivea', 'Lanbena',
       'Hada Labo', 'Himalaya', 'VT Cosmetics', 'Eaoron',
       'Vibrant Glamour', 'The Face Shop', 'Breylee',
       'sunniesfaceofficial.os', 'Huxley', 'Wis', 'telekungrawdah.os',
       'Simply K', "LIYAL'AN", 'Focallure', 'Bifesta', 'Curel', 'Sebamed',
       'KingdomCares', 'Dalton Marine Cosmetics', 'SNP'], dtype=object)

In [ ]:
len(face_care['facecare_brand'].unique())

26

##**Step 8.1: Basic Text Cleaning**
The data / text cleaning performed followed the sequence of :-
1. Convert the data type of dataframe from object to str in order to work with python string build-in functions such as string join and split and its supporting library such as library re and statistics.
2. Convert all words to lowercase to standardize all text case in order to ease the latter cleaning processing.
3. Change ampersand html representation to the word 'and'.
4. Remove all delimiters, special characters that are neither alphabeths and number.
5. Remove extra spaces between words.
6. Remove unnecessary symbol such as '.os' (represents official store) and 'rm' (Ringgit Malaysia) that hinders statistical evaluation as strings cannot work with built-in functions that are meant for data of type float. 
7. Convert numbers with symbol **k** in face care product total sold column into thousands for the purpose of statistical evaluation.
<br>
**This deals with formatting issues / fixes structural errors**


In [ ]:
def data_cleaning(dataframe, df_col1, df_col2, df_col3, df_col4, df_col5):
    df = dataframe.astype('str')
    df = df.applymap(lambda x: x.lower())
    df = df.applymap(lambda x: re.sub('&amp;', 'and', x))
    df[df_col1] = df[df_col1].map(lambda x: re.sub("'s", '', x))
    df[df_col1] = df[df_col1].map(lambda x: re.sub(r'[^a-z0-9\s]+', '', x))
    df[df_col3] = df[df_col3].map(lambda x: re.sub(r'[^a-z0-9\s]+', '', x))
    df[df_col1] = df[df_col1].map(lambda x: re.sub(',', '', x))
    df[df_col1] = df[df_col1].map(lambda x: " ".join(x.split()))
    df[df_col2] = df[df_col2].map(lambda x: re.sub('.os', '', x))
    df[df_col4] = df[df_col4].map(lambda x: re.sub('rm', '', x))
    df[df_col4] = df[df_col4].map(lambda x: statistics.median([float(element.strip()) for element in x.split('-')]))
    df[df_col5] = df[df_col5].map(lambda x: float(x.split('k')[0]) * 1000 if len(x.split('k')) > 1 else float(x.split('k')[0]))
    return df
face_care = data_cleaning(face_care, 'facecare_category', 'facecare_brand', 'facecare_name', 'facecare_price', 'facecare_total_sold')
face_care.head(10)

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0  slimmers and massagers   face factory  ...        230.000                 23.0
1         toner and mists         laikou  ...         36.990                258.0
2     face mask and packs        garnier  ...         38.250               1100.0
3  slimmers and massagers         ckeyin  ...         47.295                715.0
4                cleanser          nivea  ...         16.720               2900.0
5       essence and serum         laikou  ...         17.490               2600.0
6     face scrub and peel          nivea  ...         16.560                239.0
7     face scrub and peel          nivea  ...            NaN               2300.0
8     face mask and packs        lanbena  ...          3.590               1300.0
9     face mask and packs        lanbena  ...          5.500               5300.0

[10 rows x 5 columns]

In [ ]:
face_care['facecare_brand'].unique()

array(['face factory', 'laikou', 'garnier', 'ckeyin', 'nivea', 'lanbena',
       'hada labo', 'himalaya', 'vt metics', 'eaoron', 'vibrant glamour',
       'the face shop', 'breylee', 'sunniesfaceofficial', 'huxley', 'wis',
       'telekungrawdah', 'simply k', "liyal'an", 'focallure', 'bifesta',
       'curel', 'sebamed', 'kingdomcares', 'dalton marine metics', 'snp'],
      dtype=object)

In [ ]:
sorted(face_care['facecare_category'].unique())

['acne treatment',
 'body lotion and others',
 'body scrubs exfoliant and peel',
 'body sunscreen and after sun',
 'body wash and soap',
 'cleanser',
 'cotton balls pads and buds',
 'essence and serum',
 'eye cream and serum',
 'face mask and packs',
 'face scrub and peel',
 'facial tools',
 'foundation',
 'gift set',
 'hair and scalp treatments',
 'hair styling',
 'health accessories',
 'make up remover',
 'masks lotion cream and scrubs',
 'men grooming',
 'moisturizer and cream',
 'slimmers and massagers',
 'sunscreen and aftersun',
 'toner and mists']

##**Step 8.2: Inspection of face care product based on their category**
The face care products with categories that may not be suited with the case study are inspected to justify and ensure that the product under those categories are irrelevant to case study before removing them from the data collection.

In [ ]:
face_care[(face_care['facecare_category'] == 'body lotion and others') 
            | (face_care['facecare_category'] == 'body scrubs exfoliant and peel') 
            | (face_care['facecare_category'] == 'body sunscreen and after sun') 
            | (face_care['facecare_category'] == 'body wash and soap') 
            | (face_care['facecare_category'] == 'cotton balls pads and buds') 
            | (face_care['facecare_category'] == 'facial tools') 
            | (face_care['facecare_category'] == 'gift set') 
        ]

facecare_category  ... facecare_total_sold
42     body sunscreen and after sun  ...               282.0
50       cotton balls pads and buds  ...              2200.0
53     body sunscreen and after sun  ...               147.0
57                         gift set  ...                13.0
59     body sunscreen and after sun  ...               168.0
61   body scrubs exfoliant and peel  ...               904.0
74           body lotion and others  ...                48.0
78               body wash and soap  ...                19.0
81               body wash and soap  ...                 9.0
88                         gift set  ...                10.0
92               body wash and soap  ...                 1.0
97     body sunscreen and after sun  ...               144.0
101    body sunscreen and after sun  ...               132.0
108              body wash and soap  ...                 0.0
111              body wash and soap  ...                 3.0
113              body wash and soap  ...                 0.0
114                    facial tools  ...                24.0
118              body wash and soap  ...                 0.0
119                        gift set  ...                 0.0

[19 rows x 5 columns]

In [ ]:
face_care[(face_care['facecare_category'] == 'hair and scalp treatments')
            | (face_care['facecare_category'] == 'hair styling') 
            | (face_care['facecare_category'] == 'health accessories') 
            | (face_care['facecare_category'] == "men grooming")
            | (face_care['facecare_category'] == 'slimmers and massagers')
            | (face_care['facecare_category'] == 'sunscreen and aftersun')
        ]

facecare_category  ... facecare_total_sold
0       slimmers and massagers  ...                23.0
3       slimmers and massagers  ...               715.0
11                men grooming  ...             15600.0
15                men grooming  ...                76.0
29                men grooming  ...                43.0
31                men grooming  ...              3100.0
32                men grooming  ...              2500.0
33                men grooming  ...               122.0
35                men grooming  ...               445.0
37                men grooming  ...              1200.0
38                men grooming  ...                 7.0
39                men grooming  ...               286.0
40                men grooming  ...                60.0
44                men grooming  ...               280.0
46                men grooming  ...                99.0
48          health accessories  ...                 9.0
51                men grooming  ...              1300.0
52                men grooming  ...               236.0
54                men grooming  ...               286.0
64                men grooming  ...               138.0
65                hair styling  ...               272.0
69      sunscreen and aftersun  ...                37.0
71                men grooming  ...               403.0
75                men grooming  ...               601.0
79                men grooming  ...               404.0
82                men grooming  ...                13.0
83                men grooming  ...               652.0
86   hair and scalp treatments  ...                 0.0
89      slimmers and massagers  ...                18.0
93                men grooming  ...                 5.0
95                men grooming  ...                11.0
98                hair styling  ...                17.0
100               men grooming  ...                30.0
102               men grooming  ...                17.0
105               men grooming  ...                 5.0
106               men grooming  ...                 1.0
107               men grooming  ...               361.0
109     sunscreen and aftersun  ...                 7.0
110               men grooming  ...                 0.0
112     sunscreen and aftersun  ...                 0.0
115     sunscreen and aftersun  ...                 4.0
116     sunscreen and aftersun  ...                 1.0
117               men grooming  ...                 2.0

[43 rows x 5 columns]

##**Step 8.3: Remove Irrelevant Product**
From Step 8.2, it can be concluded that these product categories are irrelevant to the case study -> face care products. Hence, products with these categories are removed:
1. body lotion and others
2. slimmers and massagers
3. body wash and soap
4. hair and scalp treatment
5. cotton balls pads and buds
6. facial tools
7. hair styling
8. health accessories
9. body scrubs exfoliant and peel

In [ ]:
remove_irrelevant_lst = list(face_care[(face_care['facecare_category'] == 'body lotion and others') |(face_care['facecare_category'] == 'slimmers and massagers') 
            | (face_care['facecare_category'] == 'body wash and soap') | (face_care['facecare_category'] == 'hair and scalp treatments') 
            | (face_care['facecare_category'] == 'cotton balls pads and buds') | (face_care['facecare_category'] == 'facial tools') 
            | (face_care['facecare_category'] == 'hair styling') | (face_care['facecare_category'] == 'health accessories') 
            | (face_care['facecare_category'] == "body scrubs exfoliant and peel")].index)
remove_irrelevant_lst

[0, 3, 48, 50, 61, 65, 74, 78, 81, 86, 89, 92, 98, 108, 111, 113, 114, 118]

In [ ]:
face_care.drop(index=remove_irrelevant_lst, inplace=True)
face_care = face_care.reset_index(drop=True)
face_care

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0           toner and mists         laikou  ...          36.99                258.0
1       face mask and packs        garnier  ...          38.25               1100.0
2                  cleanser          nivea  ...          16.72               2900.0
3         essence and serum         laikou  ...          17.49               2600.0
4       face scrub and peel          nivea  ...          16.56                239.0
..                      ...            ...  ...            ...                  ...
97   sunscreen and aftersun  the face shop  ...          81.90                  0.0
98   sunscreen and aftersun            snp  ...          71.25                  4.0
99   sunscreen and aftersun            snp  ...          71.25                  1.0
100            men grooming         laikou  ...          49.50                  2.0
101                gift set         huxley  ...         259.90                  0.0

[102 rows x 5 columns]

##**Step 8.4: Group product category**
This part combines both categories, namely the **groups body sunscreen and after sun** and **sunscreen and aftersun** as a single category, which is **sunscreen**

In [ ]:
face_care[(face_care['facecare_category'] == 'body sunscreen and after sun') | (face_care['facecare_category'] == 'sunscreen and aftersun')]#['facecare_name'].unique()

facecare_category  ... facecare_total_sold
40  body sunscreen and after sun  ...               282.0
49  body sunscreen and after sun  ...               147.0
55  body sunscreen and after sun  ...               168.0
63        sunscreen and aftersun  ...                37.0
85  body sunscreen and after sun  ...               144.0
88  body sunscreen and after sun  ...               132.0
95        sunscreen and aftersun  ...                 7.0
97        sunscreen and aftersun  ...                 0.0
98        sunscreen and aftersun  ...                 4.0
99        sunscreen and aftersun  ...                 1.0

[10 rows x 5 columns]

In [ ]:
sunscreen_lst = list(face_care[(face_care['facecare_category'] == 'body sunscreen and after sun') | (face_care['facecare_category'] == 'sunscreen and aftersun')].index)
sunscreen_lst

[40, 49, 55, 63, 85, 88, 95, 97, 98, 99]

In [ ]:
for index_sunscreen in sunscreen_lst:
    face_care.at[index_sunscreen, 'facecare_category'] = 'sunscreen'

In [ ]:
face_care[(face_care['facecare_category'] == 'body sunscreen and after sun') | (face_care['facecare_category'] == 'sunscreen and aftersun')]

Empty DataFrame
Columns: [facecare_category, facecare_brand, facecare_name, facecare_price, facecare_total_sold]
Index: []

In [ ]:
face_care[(face_care['facecare_category'] == 'sunscreen')]

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
40         sunscreen        lanbena  ...          12.80                282.0
49         sunscreen          nivea  ...          42.42                147.0
55         sunscreen          nivea  ...          42.42                168.0
63         sunscreen  the face shop  ...          98.90                 37.0
85         sunscreen          curel  ...          74.88                144.0
88         sunscreen          curel  ...          51.44                132.0
95         sunscreen         huxley  ...          60.23                  7.0
97         sunscreen  the face shop  ...          81.90                  0.0
98         sunscreen            snp  ...          71.25                  4.0
99         sunscreen            snp  ...          71.25                  1.0

[10 rows x 5 columns]

In [ ]:
face_care.shape

(102, 5)

In [ ]:
face_care['facecare_category'].unique()

array(['toner and mists', 'face mask and packs', 'cleanser',
       'essence and serum', 'face scrub and peel', 'men grooming',
       'moisturizer and cream', 'masks lotion cream and scrubs',
       'make up remover', 'acne treatment', 'sunscreen', 'gift set',
       'foundation', 'eye cream and serum'], dtype=object)

##**Step 8.5: Handling Missing Value**
Some of the prices of face care product weren't successfully extracted, hence it is recorded as null values in the dataframe. To overcome this problem, the null values are replaced with the median price according to the product category.<br>
**Example:** Face care product at index 5 has empty value for attribute price. To overcome this problem, the median of price for products with the category same as product index 5 (which is face scrub and peel) is calculated. Then, the empty price value for index 5 is replaced with the median calculated. <br>
**Reason of applying this mentioned approach:** The majority of the prices of face care products which are of the same cateogry are almost the same.

In [ ]:
face_care['facecare_price'].isna().sum() / face_care.shape[0] *100

1.9607843137254901

In [ ]:
face_care[face_care['facecare_price'].isna()]

facecare_category  ... facecare_total_sold
5             face scrub and peel  ...              2300.0
25  masks lotion cream and scrubs  ...                86.0

[2 rows x 5 columns]

In [ ]:
category_with_nan_price = list(face_care[face_care['facecare_price'].isna() == True]['facecare_category'].unique())
category_with_nan_price

['face scrub and peel', 'masks lotion cream and scrubs']

In [ ]:
# replace missing value of product prices with the median price based on product category
def calculate_median(dataframe, column_category_lst):
    median_dictionary = dict()
    for category in column_category_lst:
        median_dictionary[category] = round(face_care[face_care['facecare_category'] == category]['facecare_price'].median(),3)
    return median_dictionary

median_dict = calculate_median(face_care, category_with_nan_price)
median_dict

{'face scrub and peel': 16.24, 'masks lotion cream and scrubs': 28.525}

In [ ]:
def fill_na(dataframe, median_dictionary):
    for key, value in median_dictionary.items():
        dataframe[dataframe['facecare_category'] == key] = dataframe[dataframe['facecare_category'] == key].fillna(value)
    return dataframe

face_care_cleaned = fill_na(face_care, median_dict)
face_care_cleaned

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0        toner and mists         laikou  ...          36.99                258.0
1    face mask and packs        garnier  ...          38.25               1100.0
2               cleanser          nivea  ...          16.72               2900.0
3      essence and serum         laikou  ...          17.49               2600.0
4    face scrub and peel          nivea  ...          16.56                239.0
..                   ...            ...  ...            ...                  ...
97             sunscreen  the face shop  ...          81.90                  0.0
98             sunscreen            snp  ...          71.25                  4.0
99             sunscreen            snp  ...          71.25                  1.0
100         men grooming         laikou  ...          49.50                  2.0
101             gift set         huxley  ...         259.90                  0.0

[102 rows x 5 columns]

In [ ]:
face_care_cleaned.isna().sum()

facecare_category      0
facecare_brand         0
facecare_name          0
facecare_price         0
facecare_total_sold    0
dtype: int64

###**SAVE THE CLEANED BUT UNTRANSFORMED DATA INTO A CSV FILE**
It will be used in Assignment 2: Data Visualization to compare and contrast the distribution of data before and after performing data transformation.

In [ ]:
face_care_cleaned.to_csv('a176297_cleaned_face_care.csv')

##**Step 8.6: Data Transformation**
The price and total sold of face care products are visualized to emphasize the need of data transformation. In this section, Z-Score data scaling technique will be used to standardize the data of column facecare_price and facecare_total_sold.

In [ ]:
face_care_cleaned = pd.read_csv('/content/a176297_cleaned_face_care.csv', index_col='Unnamed: 0')
face_care_cleaned.head()

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0      toner and mists         laikou  ...          36.99                258.0
1  face mask and packs        garnier  ...          38.25               1100.0
2             cleanser          nivea  ...          16.72               2900.0
3    essence and serum         laikou  ...          17.49               2600.0
4  face scrub and peel          nivea  ...          16.56                239.0

[5 rows x 5 columns]

In [ ]:
import plotly.express as px

def plot_histogram(dataframe, x, title, nbins=20, template='ggplot2'):
    return px.histogram(face_care_cleaned, x=x, nbins=20, template='ggplot2', title=title)

In [ ]:
plot_histogram(face_care_cleaned, x='facecare_price', title='Distribution of the prices of face care products')

In [ ]:
plot_histogram(face_care_cleaned, x='facecare_total_sold', title='Distribution of the total face care products sold')

In [ ]:
!pip install mlxtend

In [ ]:
from mlxtend.preprocessing import standardize
face_care_cleaned['facecare_price'] = standardize(face_care_cleaned['facecare_price'].to_numpy())#, columns=['facecare_price'])
face_care_cleaned['facecare_total_sold'] = standardize(face_care_cleaned['facecare_total_sold'].to_numpy())
face_care_cleaned.head()

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0      toner and mists         laikou  ...      -0.123666            -0.282947
1  face mask and packs        garnier  ...      -0.112121             0.142678
2             cleanser          nivea  ...      -0.309387             1.052564
3    essence and serum         laikou  ...      -0.302332             0.900916
4  face scrub and peel          nivea  ...      -0.310853            -0.292551

[5 rows x 5 columns]

In [ ]:
# from scipy import stats
# normalized_product_price = stats.boxcox(face_care_cleaned['facecare_price'])
# normalized_product_sales = stats.boxcox(face_care_cleaned[face_care_cleaned['facecare_total_sold']>0]['facecare_total_sold'])

# normalized_df = pd.DataFrame(normalized_product_price[0], columns=['normalized_price'])
# normalized_df['normalized_sales'] = pd.DataFrame(normalized_product_sales[0], columns=['normalized_sales'])
# normalized_df['normalized_sales'].fillna(0, inplace=True)
# normalized_df['product_category'] = face_care_cleaned['facecare_category']
# normalized_df['product_brand'] = face_care_cleaned['facecare_brand']
# normalized_df['product_name'] = face_care_cleaned['facecare_name']

# normalized_df = normalized_df.round({'normalized_price': 4, 'normalized_sales': 4})

# normalized_df

normalized_price  ...                                       product_name
0              2.7470  ...  laikou sakura toner face serum reduce spots wr...
1              2.7658  ...  garnier light complete brightening serum mask ...
2              2.2703  ...  nivea face care cleanser  hokkaido rose whip 1...
3              2.2990  ...  laikou japan sakura cleanser whitening face se...
4              2.2642  ...  nivea face care cleanser  glow raspberry scrub...
..                ...  ...                                                ...
97             3.1675  ...               natural sun eco super aqua sun cream
98             3.0975  ...  snp  uv perfect aircool sun gel  air safe sun ...
99             3.0975  ...  snp  uv perfect airlock sun fluid  air safe su...
100            2.9073  ...  laikou men  face skin care set face toner loti...
101            3.6910  ...  huxley official conditioning skincare set  ser...

[102 rows x 5 columns]

In [ ]:
normalized_df = face_care_cleaned.copy()
normalized_catdummy_df = pd.get_dummies(normalized_df['facecare_category'])
normalized_brddummy_df= pd.get_dummies(normalized_df['facecare_brand'])
cleaned_df = pd.concat([normalized_catdummy_df, normalized_brddummy_df, normalized_df['facecare_name'], normalized_df['facecare_price'], normalized_df['facecare_total_sold']], axis=1)
cleaned_df

acne treatment  cleanser  ...  facecare_price  facecare_total_sold
0                 0         0  ...       -0.123666            -0.282947
1                 0         0  ...       -0.112121             0.142678
2                 0         1  ...       -0.309387             1.052564
3                 0         0  ...       -0.302332             0.900916
4                 0         0  ...       -0.310853            -0.292551
..              ...       ...  ...             ...                  ...
97                0         0  ...        0.287817            -0.413364
98                0         0  ...        0.190238            -0.411342
99                0         0  ...        0.190238            -0.412858
100               0         0  ...       -0.009044            -0.412353
101               0         0  ...        1.918721            -0.413364

[102 rows x 37 columns]

In [ ]:
cleaned_df.isna().sum()

acne treatment                   0
cleanser                         0
essence and serum                0
eye cream and serum              0
face mask and packs              0
face scrub and peel              0
foundation                       0
gift set                         0
make up remover                  0
masks lotion cream and scrubs    0
men grooming                     0
moisturizer and cream            0
sunscreen                        0
toner and mists                  0
bifesta                          0
breylee                          0
curel                            0
dalton marine metics             0
eaoron                           0
garnier                          0
hada labo                        0
himalaya                         0
huxley                           0
laikou                           0
lanbena                          0
liyal'an                         0
nivea                            0
snp                              0
sunniesfaceofficial 

#**PART 9: SAVE CLEANED AND TRANSFORMED DATA TO A CSV FILE**
This csv file will be used in Assignment 2: Data Visualization to visualize the distribution of normalized data.

In [ ]:
cleaned_df.to_csv('a176297_cleaned_transform_one_hot_face_care.csv')
print('Cleaned, Transformed, One-Hot Encode -- Done!')

Cleaned, Transformed, One-Hot Encode -- Done!


In [ ]:
face_care_cleaned.to_csv('a176297_cleaned_transform_face_care.csv')
print('Cleaned, Transformed -- Done!')

Cleaned, Transformed -- Done!


#**PART 10: DATA VISUALIZATION**

##**Step 10.1: Import Libraries**

In [ ]:
!pip uninstall pandas-profiling
!pip install pandas-profiling

Found existing installation: pandas-profiling 3.1.0
Uninstalling pandas-profiling-3.1.0:
  Would remove:
    /usr/local/bin/pandas_profiling
    /usr/local/lib/python3.7/dist-packages/pandas_profiling-3.1.0.dist-info/*
    /usr/local/lib/python3.7/dist-packages/pandas_profiling/*
Proceed (y/n)? y
  Successfully uninstalled pandas-profiling-3.1.0
  Using cached pandas_profiling-3.1.0-py2.py3-none-any.whl (261 kB)


In [ ]:
import plotly.express as px
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.figure_factory as ff
from pandas_profiling import ProfileReport
import pandas as pd

In [ ]:
df = pd.read_csv('/content/a176297_cleaned_face_care.csv', index_col='Unnamed: 0')
df

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0        toner and mists         laikou  ...          36.99                258.0
1    face mask and packs        garnier  ...          38.25               1100.0
2               cleanser          nivea  ...          16.72               2900.0
3      essence and serum         laikou  ...          17.49               2600.0
4    face scrub and peel          nivea  ...          16.56                239.0
..                   ...            ...  ...            ...                  ...
97             sunscreen  the face shop  ...          81.90                  0.0
98             sunscreen            snp  ...          71.25                  4.0
99             sunscreen            snp  ...          71.25                  1.0
100         men grooming         laikou  ...          49.50                  2.0
101             gift set         huxley  ...         259.90                  0.0

[102 rows x 5 columns]

In [ ]:
df.describe()

facecare_price  facecare_total_sold
count      102.000000           102.000000
mean        50.487108           817.745098
std        109.680912          1988.039194
min          2.950000             0.000000
25%         15.920000            11.000000
50%         24.725000           135.000000
75%         49.340000           592.000000
max        999.000000         15600.000000

In [ ]:
df_norm = pd.read_csv('/content/a176297_cleaned_transform_face_care.csv', index_col='Unnamed: 0')
df_norm

facecare_category facecare_brand  ... facecare_price  facecare_total_sold
0        toner and mists         laikou  ...      -0.123666            -0.282947
1    face mask and packs        garnier  ...      -0.112121             0.142678
2               cleanser          nivea  ...      -0.309387             1.052564
3      essence and serum         laikou  ...      -0.302332             0.900916
4    face scrub and peel          nivea  ...      -0.310853            -0.292551
..                   ...            ...  ...            ...                  ...
97             sunscreen  the face shop  ...       0.287817            -0.413364
98             sunscreen            snp  ...       0.190238            -0.411342
99             sunscreen            snp  ...       0.190238            -0.412858
100         men grooming         laikou  ...      -0.009044            -0.412353
101             gift set         huxley  ...       1.918721            -0.413364

[102 rows x 5 columns]

In [ ]:
df_norm.describe()

facecare_price  facecare_total_sold
count    1.020000e+02         1.020000e+02
mean     7.401487e-17        -4.625929e-17
std      1.004938e+00         1.004938e+00
min     -4.355531e-01        -4.133638e-01
25%     -3.167170e-01        -4.078034e-01
50%     -2.360422e-01        -3.451223e-01
75%     -1.051024e-02        -1.141124e-01
max      8.690636e+00         7.472315e+00

##**Step 10.2: Visualization using Histogram and BoxPlot on Face Care Product Price**

###**Unnormalized Data**

In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

fig = make_subplots(rows=1, cols=2)

fig.add_trace(
    go.Box(y=df['facecare_price'].tolist(), boxpoints='outliers'),
    row=1, col=1, 
)

fig.add_trace(
    go.Histogram(x = df['facecare_price'].tolist()),
    row=1, col=2,
    
)

fig.update_layout(height=600, width=800, title_text="Distribution of the price of face care products", title_x=0.5)
fig['layout']['xaxis']['title']= ' '
fig['layout']['xaxis2']['title']='Price (RM)'
fig['layout']['yaxis']['title']='Price (RM)'
fig['layout']['yaxis2']['title']='Count'
fig.show()

###**Normalized Data**

In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

fig = make_subplots(rows=1, cols=2)

fig.add_trace(
    go.Box(y=df_norm['facecare_price'].tolist(), boxpoints='outliers'),
    row=1, col=1, 
)

fig.add_trace(
    go.Histogram(x = df_norm['facecare_price'].tolist()),
    row=1, col=2,
)

fig.update_layout(height=600, width=800, title_text="Distribution of the scaled price of face care products", title_x=0.5)
fig['layout']['xaxis']['title']= ' '
fig['layout']['xaxis2']['title']='Price (RM)'
fig['layout']['yaxis']['title']='Price (RM)'
fig['layout']['yaxis2']['title']='Count'
fig.show()

In [ ]:
df_norm.skew(axis=0)

facecare_price         6.948274
facecare_total_sold    5.008015
dtype: float64

###**Data Description on Face Care Product Price:**
1. **BoxPlot** <br>
    * From the boxplot, it is observed that there are 6 outliers in the distribution of face care product price data. All outliers are located beyond the maximum observation upper fence / upper extreme that has the value of 0.4490748.
    * The maximum value for the face care product price is 8.690636 while the minimum value is -0.4355531.
    * The first quartile (Q1) is -0.316717, median (Q2) is -0.2360422 and thrid quartile (Q3) is -0.009044258.
    * From the boxplot, it can be concluded that the distribution of price for face care products is positively skewed (+6.948274) as Q2-Q1 < Q3-Q2. Besides, positive skewness indicates that the value of low range prices are close as compared to high range prices that has a wider distribution.

2. **Histogram**
    * There are outliers as shown in the histogram, indicated by the presence of isolated histogram bars or bars that are further away from the majority.
    * From the shape of the histogram, it is observed that the distribution of face care product price is positively skewed. 

---
**Note: The product with price RM999 is inspected and it is found that this product is no longer for sales but the seller not only did not remove the product from Shopee, but it sets the price for this product as RM999**


##**Step 10.3: Visualization using Scatter Plot on Face Care Product Price and Total Sold**

###**Unnormalized Data**

In [ ]:
fig_scatter = px.scatter(df, x='facecare_price', y='facecare_total_sold', color='facecare_category', template='ggplot2', hover_name=df['facecare_name'], opacity=0.6)
fig_scatter.show()

###**Normalized Data**

In [ ]:
fig_scatter_norm = px.scatter(df_norm, x='facecare_price', y='facecare_total_sold', template='ggplot2', color='facecare_category', hover_name=df_norm['facecare_name'], opacity=0.6)
fig_scatter_norm.show()

In [ ]:
px.scatter_matrix(df_norm[['facecare_price', 'facecare_total_sold']])

###**Description on the Distribution of data for Face Care Product Price and its Total Quantity Sold:**
The general pattern of the distribution of the scatter plot is that the lower the price of face care products, the higher the total quantity of face care products sold out. Therefore, the total sold of face care product is inversely proportional to the face care product price. For instance, by comparing the product "huxley official exclusive power skincare set face toner essence eye cream" and "hada labo premium whitening facewash 100g", the huxley face care product has a lower sales as it has a higher price while the hada labo facewash achieved the highest number of sales as its price is considerably low.<br><br>
Moreover, the scatter plot distribution has outliers as shown by data points that are far away from the crowd. This is explained by the example of "huxley official exclusive power skincare set face toner essence eye cream" and "hada labo premium whitening facewash 100g" again. Even though both products have a higher price and sales respectively, but when both of them are to be compared to the rest of the face care products, they are considered as outliers. This is because the majority of the face care products doesn't have price and total number of sales as high as "huxley official exclusive power skincare set face toner essence eye cream" and "hada labo premium whitening facewash 100g" respectively. <br><br>

##**Step 10.4: Data Profiling**
It is an exploratory data analysis via visualization that consists of statistical assessment of the collection of data scraped. The profile report is downloaded as a html page. From the profile report generated, there is 1 row of duplicated data in the dataset. The dupicated data will be deleted before feeding the data to machine learning algorithms.

In [ ]:
prof = ProfileReport(df)
prof.to_file(output_file='a176297_output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
prof = ProfileReport(df_norm)
prof.to_file(output_file='a176297_norm_output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

##**Extra Visualization**

###**1.1 Visualization of the total number of face care products according to the category and brand of the face care product.**

In [ ]:
fig_brand = px.histogram(df, y="facecare_category", title='The count of different face care product category', template='ggplot2', 
                         orientation='h', color='facecare_brand', opacity=0.8, color_discrete_sequence=px.colors.diverging.curl)
fig_brand.show()

**From the bar chart above, the majority of the face care products has category of men grooming and the brand hada labo constitutes for the most number of products for this category. In contrast, there is only one product in each of the category eye cream and serum (brand: Dalton Marine Metics), and foundation (brand: Huxley).** 

###**1.2 Visualization of the total number of face care products according to the face care product brand.**

In [ ]:
unique_brand = list(df['facecare_brand'].unique())
array = []
for brand in unique_brand:
    array.append([brand, df[df['facecare_brand']==brand]['facecare_total_sold'].sum()])
modified_df = pd.DataFrame(array, columns=['facecare_brand', 'facecare_total_sold'])
modified_df

facecare_brand  facecare_total_sold
0                 laikou               7820.0
1                garnier               1100.0
2                  nivea              22649.0
3                lanbena              15572.0
4              hada labo              16237.0
5               himalaya               5029.0
6              vt metics                419.0
7                 eaoron               1900.0
8        vibrant glamour               2585.0
9          the face shop                177.0
10               breylee               8700.0
11   sunniesfaceofficial                122.0
12                huxley                 63.0
13                   wis                 99.0
14        telekungrawdah                207.0
15              liyal'an                  0.0
16               bifesta                444.0
17                 curel                282.0
18  dalton marine metics                  0.0
19                   snp                  5.0

In [ ]:
fig_brand = px.bar(modified_df, x="facecare_brand", y="facecare_total_sold", title='Total number of face care product sold by each face care product brand', template='ggplot2', 
                    opacity=0.8, color_discrete_sequence=px.colors.diverging.curl)
fig_brand.show()

**From the bar chart above, Nivea made the highest number of sales while Dalton Marine Metics and Liyal'an have not made any sales yet.** 

**END OF PROJECT**